### ISSUE TITLE
Bug in test function: Incorrect slicing on batch_y due to mismatched dimension semantics

### ISSUE THREAD

**Issue by Mona-chaser** [BennyTMT/LLMsForTimeSeries#10]
Hi, thank you very much for your excellent work and for sharing the code with the community. It has been very helpful for my research and understanding.

While reading through the code, I noticed a potential issue in the test function that might affect the correctness of the evaluation. Specifically, in the following lines:

**Issue location:** [test function in tools.py](https://github.com/BennyTMT/LLMsForTimeSeries/blob/65a8897ce06abc9db4aa60ae7eb677661a58c48a/PAttn/utils/tools.py#L432)

```
 def test(model, test_data, test_loader, args, device, itr):
    preds = []
    trues = []
    prevs = []
    model.eval()
    with torch.no_grad():
        for i, (batch_x, batch_y, batch_x_mark, batch_y_mark) in tqdm(enumerate(test_loader)):
            
        
            batch_x = batch_x.float().to(device)
            batch_y = batch_y.float().to(device)
            
            outputs = model(batch_x[:, -args.seq_len:, :])
            
            # encoder - decoder
            outputs = outputs[:, -args.pred_len:, :]
            batch_y = batch_y[:, -args.pred_len:, :].to(device)


            pred = outputs.detach().cpu().numpy()
            true = batch_y.detach().cpu().numpy()
            prev = batch_x[:, -args.seq_len:, :].cpu().numpy()
            
            preds.append(pred)
            trues.append(true)
            prevs.append(prev)
```
From the debug results using ILI dataset, it's shown that `batch_y` has the shape `(batch_size, dim, pred_len)=(8, 7, 24)` — where `dim` represents the feature dimensions, not the time steps. Your comments in the training code also seem to confirm this observation:

**Location:** [train data in main.py](https://github.com/BennyTMT/LLMsForTimeSeries/blob/65a8897ce06abc9db4aa60ae7eb677661a58c48a/PAttn/main.py#L180C17-L181C70)

```
        for i, (batch_x, batch_y, batch_x_mark, batch_y_mark) in tqdm(enumerate(train_loader)):
            
            if is_first : 
                # torch.Size([512, 7, 96]) torch.Size([512, 7, 96])
                print(args.data_path, batch_x.shape  , batch_y.shape)
                is_first=False
```
It may mean that for ETTh1, batch_size=512, dim=7, pred_len=96, seq_len=96, If I understand correctly?

Therefore, slicing batch_y using [:, -args.pred_len:, :] is likely incorrect because the second dimension is not the time axis. The slicing should likely target the third dimension (i.e., the prediction length).

I would like to confirm whether this is indeed an issue, and if so, whether it could impact the results presented in the paper.

Thanks again for your contributions!

**Comment by BennyTMT** [comment]
Thanks so much for your interest and reminder. PAttn was built on the One-Fits-All repo, and we did notice their slicing behavior.

I’ve tested it—this specific slicing in the test phase isn't necessary and generally doesn’t affect the results, 
```
outputs = outputs[:, -args.pred_len:, :]
batch_y = batch_y[:, -args.pred_len:, :].to(device)
```

as `pred_len` is often larger than `dim`, so the returned array remains unchanged. **That said, we truly appreciate your observation—we’ve updated the code and apologize for any confusion caused.**

There might be an impact on the `Traffic` dataset, since it has a large number of channels (862). PAttn got worst performance on Traffic, so we’ll run additional tests to see if this is related. **Thanks again!**

**Comment by Mona-chaser** [comment]
Hi BennyTMT,

Thank you for your thoughtful response and for taking the time to update the code. I completely agree with your point that for lower-dimensional datasets, such as ILI, the slicing might not have a noticeable impact. However, as you pointed out, the issue arises when pred_len < dim. For higher-dimensional datasets like Electricity （dim=321, pred_len=96）or Traffic, where the number of channels is larger, this could indeed affect performance.

I also noticed the same issue in the **One-Fits-All** implementation in your repo.
https://github.com/BennyTMT/LLMsForTimeSeries/blob/63748dcc1ac9a3bac6045ff2bf562a3e2a9126a6/OFA/utils/tools.py#L307C17-L307C68
```
                outputs = model(batch_x[:, -args.seq_len:, :], itr)
                
                # encoder - decoder
                outputs = outputs[:, -args.pred_len:, :]
                batch_y = batch_y[:, -args.pred_len:, :].to(device)
```

I appreciate your attention to this matter and the prompt updates to the code. Your contributions and the continued improvements are very valuable!

**Comment by BennyTMT** [comment]
Thank you for following up.

After fixing the code yesterday, I retrained and tested the performance of PAttn on the `Traffic` dataset, and the results is below,
```
	   MAE    MSE
96       0.264    0.396
192      0.267    0.404
336      0.271    0.414
720      0.290    0.450
```
 were consistent with before,

<img width="164" alt="Image" src="https://github.com/user-attachments/assets/e0079381-c8c0-40b1-8ca7-41d0fcbb09ec" />

The One-Fits-All approach handled this correctly, which was the original design. In developing simple methods like PAttn, we made adjustments to how channels were processed,

```
x = torch.tensor(seq_x, dtype=torch.float).transpose(1, 0)  # [c, seq_len]
y = torch.tensor(seq_y, dtype=torch.float).transpose(1, 0)  # [c, pred_len]
```
The bug may be due to us developing different "simple models" within the same repo and have caused some confusion during code cleanup.

**Comment by BennyTMT** [comment]
Another dataset is similar:

After fixing the code, I retrained and tested PAttn on the `Electricity` today

```
       MAE     MSE
96    0.227    0.135
192   0.239    0.149
336   0.255    0.166
720   0.287    0.203   
```
were consistent with before,

![Image](https://github.com/user-attachments/assets/648d6801-4f84-4d2e-825d-e56dbb4181d2)

**Comment by Mona-chaser** [comment]
Ok thanks! I will close this issue.
